### DISCUSSION TITLE
CycloneDX report doesn't seem to have the correct license information

### DISCUSSION THREAD

**Discussion by jerbob92** [aquasecurity/trivy#6761] [Bugs]
### Description

In an earlier verson of the CycloneDX output format, it still had the `expression` field under `licenses`, and Trivy would generate a SPDX Expression for the found licenses. It looks like Trivy now puts every separate license in it's own license object:

```
    {
      "bom-ref": "pkg:golang/ariga.io/atlas/cmd/atlas@0.13.1",
      "type": "library",
      "name": "ariga.io/atlas/cmd/atlas",
      "version": "0.13.1",
      "licenses": [
        {
          "license": {
            "name": "Apache-2.0"
          }
        }
      ]
```

The field that's being used here is`name`, but it seems to contain a `SPDX License ID`, and not the license name, because that would probably be `Apache License 2.0`. Shouldn't the value that goes into `name` now go into `id`?

### Desired Behavior

I expected the output to have a valid SPDX expression or have the SPDX ID in the `id` property.

### Actual Behavior

I got no `expression` and got the ID of the license in the `name` field.

### Reproduction Steps

Run 

```bash
trivy fs --format cyclonedx --scanners license --license-full --output result.json .
```

In a project that has dependencies

### Target

Filesystem

### Scanner

License

### Output Format

CycloneDX

### Mode

Standalone

### Debug Output

Probably not relevant?

### Operating System

Ubuntu 20.04

### Version

```bash
Version: 0.51.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-05-23 18:11:51.32771239 +0000 UTC
  NextUpdate: 2024-05-24 00:11:51.32771225 +0000 UTC
  DownloadedAt: 2024-05-23 18:21:24.571822722 +0000 UTC
Check Bundle:
  Digest: sha256:2f95caeff50df1f00efdf5cb619c3b5488bbbb9bb08ef0890f52352464d35c79
  DownloadedAt: 2023-04-11 07:47:52.913947777 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by jerbob92** [comment]
Actually it looks like the software I was importing this into is working fine with it being in the name, it's just that locally I have Go installed and the packages downloaded locally, but in the prod runtime I don't, so it was not resolving the licenses at all...

**Reply by keremsan** [reply]
@jerbob92 I don't quite understand, Is Trivy able to put SPDX license expression into cyclonedx sbom? I'm asking because, I couldn't generate CycloneDX sbom with license expressions. It seems to put every license into a licenses array  (I'm coming from [dependency-track discussion](https://github.com/DependencyTrack/dependency-track/issues/170#issuecomment-2127685956)

**Reply by jerbob92** [reply]
@keremsan I think Trivy changed their license reporting at some point, afaik it were SPDX expressions before. I don't use it anymore because I did not want to download all dependencies to be able to know the licenses so I currently post-process the Trivy SBOM's to collect license information from other sources.

However, I dont think dependency-track is able to handle SPDX from CycloneDX properly right now.

**Reply by keremsan** [reply]
> However, I dont think dependency-track is able to handle SPDX from CycloneDX properly right now.

Yes, you're right. It's not. But It does parse "expression" keyword in license array and put it's value to SPDX expression. When you want to play with policy violation, It works with SPDX expression.  

So I think It would be better to find another tool to generate CycloneDX SBOM which will has expression.

**Reply by jerbob92** [reply]
@keremsan 
> It's not. But It does parse "expression" keyword in license array and put it's value to SPDX expression. When you want to play with policy violation, It works with SPDX expression.

Hmm, maybe they fixed that then, I could not get it to work, when I created an SBOM with expression it was seen as an invalid file. 

Do you have an example SBOM to demonstrate? Maybe I can contribute this as change to Trivy.

**Reply by keremsan** [reply]
@jerbob92 
That's another issue related to the SBOM generated by Trivy. I need to disable bom validation otherwise validation is failed. However, when I upload with validation disabled, I'm able to upload SBOM properly and I receive the expected results. (all components, vulnerabilities, licenses, etc.) 

You can try modifying a license key (name within license array) within an SBOM generated by Trivy by replacing it with "expression" and inserting an SPDX expression. Example:
```
      "licenses": [
        {
          "license": {
            "expression": "EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0"
          }
        }
      ]
```

**Reply by jerbob92** [reply]
@keremsan Is that SBOM validation issue related to the license expression then? I can upload SBOM's created by Trivy just fine, I don't have validations disabled.

**Reply by jerbob92** [reply]
@keremsan I just tested this out and it does not work. When I have validation enabled it reports that the schema doesn't match, when I disable the validation it does import the SBOM but the license that was turned into an expression is not imported. 

Trivy: `v0.59.1` (Produces CycloneDX 1.6 SBOM)
DependencyTrack: `v4.12.3`

It does work with this sample and validation disabled:

```
 "licenses": [
        {
          "expression": "EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0"
        }
      ]
```

:partying_face: 

I do wonder why it won't validate, since this is valid CycloneDX 1.6.
